sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -299,6 +300,26 @@ object ReorderAssociativeOperator extends Rule[LogicalPlan] {
 }
 
 
+/**
+ * The "ArrayContains" node checks if a given value is present in a given array,
+ * being similar to the "IN" predicate.
+ *
+ * This optimization rule replaces the "ArrayContains" nodes working on literal
+ * lists with "InSet" nodes in the expression.
+ */
+object ReplaceArrayContainsWithInSet extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan transformAllExpressions  {
+      // In Catalyst, arrays evaluated as NULL values have "arrayParam.value = NULL" set,
+      // so transforming its value to a set of objects will fail.
+      case ArrayContains(arrayParam: Literal, col)
+        if arrayParam.value != null =>
+          InSet(col, arrayParam.value.asInstanceOf[GenericArrayData].array.toSet)
```

Comment:
It's risky to assume the value is `GenericArrayData`. We can do `arrayParam.value.toSeq(col.dataType).toSet`.